tools/migration-unified-annotation-versioning/migration.py

```diff
@@ -0,0 +1,529 @@
+            if batch_start > next_version:
+                continue
+            update_groups = self.get_update_batch(tracing_or_mapping_id, collection, batch_start, batch_end)
+            for version, update_group in reversed(update_groups):
+                if version > next_version:
+                    continue
+                update_group, timestamp, revert_source_version = self.process_update_group(tracing_or_mapping_id, layer_type, update_group, json_encoder, json_decoder)
+                if revert_source_version is not None:
+                    next_version = revert_source_version
+                else:
+                    next_version -= 1
+                if revert_source_version is None:  # skip the revert itself too, since we’re ironing them out
+                    updates_for_layer.append((timestamp, version, update_group))
+        updates_for_layer.reverse()
+        return updates_for_layer
+
+    def migrate_updates(self, annotation, mapping_id_map: MappingIdMap) -> LayerVersionMapping:
+        all_update_groups = []
+        json_encoder = msgspec.json.Encoder()
+        json_decoder = msgspec.json.Decoder()
+        layers = list(annotation["layers"].items())
+        for tracing_id, layer_type in layers:
+            collection = self.update_collection_for_layer_type(layer_type)
+            all_update_groups.append(self.fetch_updates(tracing_id, layer_type, collection, json_encoder=json_encoder, json_decoder=json_decoder))
+            if tracing_id in mapping_id_map:
+                editable_mapping_id = mapping_id_map[tracing_id]
+                all_update_groups.append(self.fetch_updates(editable_mapping_id, "editableMapping", "editableMappingUpdates", json_encoder=json_encoder, json_decoder=json_decoder))
+
+        unified_version = 0
+        version_mapping = {}
+        for tracing_id, _ in layers:
+            version_mapping[tracing_id] = {0: 0} # We always want to keep the initial version 0 of all layers, even if there are no updates at all.
+
+        queue = []
+        for i, update_groups_for_layer in enumerate(all_update_groups):
+            if update_groups_for_layer:
+                # The priority queue sorts tupley lexicographically, so timestamp is the main sorting key here
+                heapq.heappush(queue, (update_groups_for_layer[0], i, 0))
+        while queue:
+            value, layer_index, element_index = heapq.heappop(queue)
+            timestamp, version, update_group = value
+            tracing_id = layers[layer_index][0]
+
+            unified_version += 1
+            version_mapping[tracing_id][version] = unified_version
+            self.save_update_group(annotation['_id'], unified_version, update_group)
+
+            if element_index + 1 < len(all_update_groups[layer_index]):
+                next_element = all_update_groups[layer_index][element_index + 1]
+                heapq.heappush(queue, (next_element, layer_index, element_index + 1))
```

Comment:
I suppose this is doing the merge sort? Not sure what is going on exactly, though.